packages/engine/state/collectible/index.ts

```diff
@@ -13,20 +14,61 @@ export const collectibleActions = {
 			collectionState.map.set(item._id, item);
 		}
 	},
-	updateCollectibleAmount: (id: string, amount: number) => {
+	setStorageCollectible: async (id: string, item: CollectibleDocument) => {
+		await modules.storage.upsert<CollectibleDocument>(id, async () => item);
+	},
+	setStorageCollection: async (id: string, item: CollectionDocument) => {
+		await modules.storage.upsert<CollectionDocument>(id, async () => item);
+	},
+	updateCollectibleAmount: async (id: string, amount: number) => {
 		const collectible = collectibleState.map.get(id);
 
-		if (collectible) {
-			const collection = collectionState.map.get(collectible.collectionId);
-			if (collection) {
-				collection.count += amount - collectible.account.amount;
-			}
-			collectible.account.amount = amount;
+		if (!collectible) return false;
+
+		const result = await modules.storage.upsert<CollectibleDocument>(
+			id,
+			async (prevDoc) => {
+				prevDoc.account.amount = amount;
+
+				return prevDoc;
+			},
+		);
+
+		return result.ok;
+	},
+	updateCollectionAmount: async (id: string, count: number) => {
+		const result = await modules.storage.upsert<CollectionDocument>(
+			id,
+			async (prevDoc) => {
+				prevDoc.count = count;
+
+				return prevDoc;
+			},
+		);
+
+		return result.ok;
+	},
+	removeCollectibleDoc: async (id: string) => {
+		const collectible = await modules.storage.safeGet<CollectibleDocument>(id);
+		const collection = await modules.storage.safeGet<CollectionDocument>(
+			collectible!.collectionId,
+		);
+		const collectionCount = collection!.count - 1;
+
+		if (collectionCount === 0) {
+			await modules.storage.removeDoc(collection!._id);
+		} else {
+			await modules.storage.upsert<CollectionDocument>(
+				collection!._id,
+				async (prevDoc) => {
+					prevDoc.count -= 1;
```

Comment:
these functions are in data storage level, so I prefer putting them into storage package